clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -172,606 +69,41 @@ public boolean hasAvailableFetches() {
-        String errorMessage = "Fetch position " + fetchPosition + " is out of range for partition " + topicPartition;
-        if (subscriptions.hasDefaultOffsetResetPolicy()) {
-            log.info("{}, resetting offset", errorMessage);
-            subscriptions.requestOffsetReset(topicPartition);
-        } else {
-            log.info("{}, raising error to the application since no reset policy is configured", errorMessage);
-            throw new OffsetOutOfRangeException(errorMessage,
-                Collections.singletonMap(topicPartition, fetchPosition.offset));
-        }
-    }
-
-    /**
-     * Clear the buffered data which are not a part of newly assigned partitions
-     *
-     * @param assignedPartitions  newly assigned {@link TopicPartition}
-     */
-    public void clearBufferedDataForUnassignedPartitions(Collection<TopicPartition> assignedPartitions) {
-        Iterator<CompletedFetch<K, V>> completedFetchesItr = completedFetches.iterator();
-        while (completedFetchesItr.hasNext()) {
-            CompletedFetch<K, V> records = completedFetchesItr.next();
-            TopicPartition tp = records.partition;
-            if (!assignedPartitions.contains(tp)) {
-                records.drain();
-                completedFetchesItr.remove();
-            }
-        }
-
-        if (nextInLineFetch != null && !assignedPartitions.contains(nextInLineFetch.partition)) {
-            nextInLineFetch.drain();
-            nextInLineFetch = null;
+            final RequestFuture<ClientResponse> future = client.send(fetchTarget, request);
+            future.addListener(listener);
         }
-    }
-
-    /**
-     * Clear the buffered data which are not a part of newly assigned topics
-     *
-     * @param assignedTopics  newly assigned topics
-     */
-    public void clearBufferedDataForUnassignedTopics(Collection<String> assignedTopics) {
-        Set<TopicPartition> currentTopicPartitions = new HashSet<>();
-        for (TopicPartition tp : subscriptions.assignedPartitions()) {
-            if (assignedTopics.contains(tp.topic())) {
-                currentTopicPartitions.add(tp);
-            }
-        }
-        clearBufferedDataForUnassignedPartitions(currentTopicPartitions);
-    }
 
-    // Visible for testing
-    protected FetchSessionHandler sessionHandler(int node) {
-        return sessionHandlers.get(node);
+        return fetchRequestMap.size();
     }
 
-    public static Sensor throttleTimeSensor(Metrics metrics, FetchMetricsRegistry metricsRegistry) {
-        Sensor fetchThrottleTimeSensor = metrics.sensor("fetch-throttle-time");
-        fetchThrottleTimeSensor.add(metrics.metricInstance(metricsRegistry.fetchThrottleTimeAvg), new Avg());
-
-        fetchThrottleTimeSensor.add(metrics.metricInstance(metricsRegistry.fetchThrottleTimeMax), new Max());
-
-        return fetchThrottleTimeSensor;
+    @Override
+    protected synchronized void handleFetchResponse(Node fetchTarget,
```

Comment:
What about just letting the super's func to be synchronized? Any concerns for that option?

Reply:
Ditto below.

Reply:
The existing `Fetcher` needs this synchronization because it's potentially called from the application thread and the heartbeat thread. However, there's going to be another subclass of `AbstractFetch` that will be used in the new consumer threading refactor, and explicitly designed such that both sending requests and handling responses will be done in the single background thread.

Granted, I don't think that it will cause any issues if the methods are synchronized. I was attempting to make the distinction that the `Fetcher` needed the synchronization.

Another option would be to remove those overrides and just wrap the calls to `handleFetchResponse` from `onSuccess` and `onFailure` inside synchronized blocks (which is what the current `Fetcher` code does today).

Reply:
I see, that makes sense, let's keep it as-is in this PR.